### PR TITLE
Blog-app: Processing data

### PR DESCRIPTION
**Pull request by GedeonTS**
# Blog-app: Processing data

In this PR I:

### Created models
-  Create model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd.png).
-  Set up associations between models.

### Used models to insert data by
- Creating at least one user 
- Creating at least 4 posts written by one of the users created previously
- Creating at least 6 posts comments for one of the posts 
- Using other [CRUD methods](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/lessons/active_record_crud.md) to find, update and delete entities.

### Created custom methods
-  Users
    - A method that returns the 3 most recent posts for a given user.
-  Posts
    - A method that updates the posts counter for a user.
    - A method which returns the 5 most recent comments for a given post.
-  Comments
    - A method that updates the comments counter for a post.
-  Likes
    - A method that updates the likes counter for a post.
- Went to `rails c` and checked if my methods are working.

### CODE REVIEW

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Gedeon, 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project, whereby the ReadMe is incomplete.

- [ ] Kindly add a detailed ReadMe to reflect your project. You can use [this template](https://github.com/microverseinc/readme-template/blob/master/README.md).

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class User < ApplicationRecord
+  has_many :comments
+  has_many :likes
+  has_many :posts
```
Comment (rloterh) [inline app/models/user.rb]: - [ ] As per the given _ERD_, the _User_ **association** should have _author_id_ referenced as **foreign key** in _Post_, _Comment,_ and _Like_ models. Kindly ensure to improve this aspect, as can be seen below:
```ruby
has_many :posts, foreign_key: 'author_id'
has_many :comments, foreign_key: 'author_id'
has_many :likes, foreign_key: 'author_id'
```

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class Post < ApplicationRecord
+  belongs_to :user
```
Comment (rloterh) [inline app/models/post.rb]: - [ ] As per the given _ERD_, the _Post_ **association** should have _Post_ belong to _User_ **through** _Author_, an alias of _User_, like below:
`belongs_to :author, class_name: 'User'` 

    Kindly ensure to improve upon this aspect.

File: app/models/like.rb [hunk 3]
```diff
@@ -0,0 +1,4 @@
+class Like < ApplicationRecord
+  belongs_to :user
```
Comment (rloterh) [inline app/models/like.rb]: - [ ] As per the given _ERD_, the _Like_ **association** should have _Like_ belong to _User_ **through** _Author_, an alias of _User_, like below:
`belongs_to :author, class_name: 'User'` 

    Kindly ensure to improve upon this aspect.

File: app/models/comment.rb [hunk 4]
```diff
@@ -0,0 +1,4 @@
+class Comment < ApplicationRecord
+  belongs_to :user
```
Comment (rloterh) [inline app/models/comment.rb]: - [ ] As per the given _ERD_, the _Comment_ **association** should have _Comment_ belong to _User_ **through** _Author_, an alias of _User_, like below:
`belongs_to :author, class_name: 'User'` 

    Kindly ensure to improve upon this aspect.

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class User < ApplicationRecord
```
Comment (rloterh) [inline app/models/user.rb:1]: - [ ] As part of this module requirement, _User_ model should have a method that returns **3 most recent posts**. This should be similar to:
```ruby
# app/models/user.rb
...

def recent_post
  posts.order('created_at Desc').limit(3)
end
```

File: app/models/post.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class Post < ApplicationRecord
```
Comment (rloterh) [inline app/models/post.rb:1]: - [ ] As part of this module requirement, _Post_ model should have a method that updates the **posts counter** for a user. This should be similar to:
```ruby
# app/models/post.rb
...

def update_post_counter
  author.increment!(:posts_counter)
end
```

Comment (rloterh) [inline app/models/post.rb:1]: - [ ] As part of this module requirement, _Post_ model should have a method that returns the **5 most recent comment**s for a given post. This should be similar to:

```ruby
# app/models/post.rb
...

def recent_comments
  comments.order('created_at Desc').limit(5)
end

File: app/models/comment.rb [hunk 7]
```diff
@@ -0,0 +1,4 @@
+class Comment < ApplicationRecord
```
Comment (rloterh) [inline app/models/comment.rb:1]: - [ ] As part of this module requirement, _Comments_ model should have a method that **updates the comments counter** for a post. This should be similar to:

```ruby
# app/models/comment.rb
...

def update_comments_counter
  post.increment!(:comments_counter)
end
```

File: app/models/like.rb [hunk 8]
```diff
@@ -0,0 +1,4 @@
+class Like < ApplicationRecord
```
Comment (rloterh) [inline app/models/like.rb:1]: - [ ] As part of this module requirement, _Like_ model should have a method that **updates the likes counter for a post. This should be similar to:

```ruby
# app/models/like.rb
...

def update_likes_counter
  post.increment!(:likes_counter)
end
```

File: app/models/post.rb (hunk 6 above)
Comment (rloterh) [inline app/models/post.rb:1]: - [ ] OPTIONAL
It is a nice idea to use one of the Rails Active Callbacks `after_save` to call the `update_user_posts_counter` method every time a new post has been saved to database, then make the `update_user_posts_counter` method private, similar to:

```ruby
 # app/models/post.rb
  after_save :update_post_counter

  private

  def update_post_counter
    author.increment!(:posts_counter)
  end
```

File: app/models/like.rb (hunk 8 above)
Comment (rloterh) [inline app/models/like.rb:1]: - [ ] OPTIONAL
It is a nice idea to use one of the Rails Active Callbacks `after_save` to call the `update_post_likes_counter method every time a new like has been saved to database, then make the `update_post_likes_counter` method private, similar to:

```ruby
 # app/models/likes.rb
  after_save :update_like_counter

  private

  def update_like_counter
    author.increment!(:likes_counter)
  end
```

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Gedeon, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 🚦 🚦 🚦  

## Highlights 👍  
- Descriptive Pull Request


## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

_Kindly check the comments marked **[OPTIONAL]** under the review._

You can also consider:
- It's nice to add a description to your introduction section of your ReadMe. This will make your ReadMe more professional.

<br>
Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by RolandM99 — APPROVED** [review]
<h1>Status: Approve 🥇 ✔</h1>
Hi @GedeonTS, 

### Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
## Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**